test-case/check-audio-equalizer.sh

```diff
@@ -68,77 +68,104 @@ current_dir=$(dirname ${BASH_SOURCE[0]})
+            dloge "IIR flter coeff list error!"
             exit 1
-    fi
-
-    declare -a IIRList=($(ls -d ${current_dir}/eqctl/eq_iir_*.txt))
-    nIIRList=${#IIRList[@]}
-    dlogi "IIR list $nIIRList ${IIRList[*]}"
-    declare -a FIRList=($(ls -d ${current_dir}/eqctl/eq_fir_*.txt))
-    nFIRList=${#FIRList[@]}
-    dlogi "FIR list $nFIRList ${FIRList[*]}"
-    if [ $nIIRList ==  0 ] || [ $nFIRList ==  0 ]; then
-        dloge "IIR or FIR flter coeff list error!"
-	exit 1
-    fi
+        fi
 
-    for i in $(seq 1 $loop_cnt)
-    do
-        dlogi "[$i/$loop_cnt] 1. Test IIR config list, IIR amixer control id=$IIRid"
-        for config in ${IIRList[@]}; do
-            func_test_eq $IIRid $current_dir/$config
-            result=$?
-            if [[ $result -ne 0 ]]; then
-                dloge "Failed at $config"
-                let failed_cnt++
+        for i in $(seq 1 $loop_cnt)
+        do
+            dlogi "[$i/$loop_cnt] 1.3 Test IIR config list, IIR amixer control id=$IIRid"
+            for config in ${IIRList[@]}; do
+                func_test_eq $IIRid $current_dir/$config
+                result=$?
+                if [[ $result -ne 0 ]]; then
+                    dloge "Failed at $config"
+                    let failed_cnt++
+                fi
+            done
+
+            dlogi "IIR test done: failed_cnt is $failed_cnt"
+            if [ $failed_cnt -gt 0 ]; then
+                exit 1
             fi
         done
+    fi # end of IIR Test
 
-        dlogi "[$i/$loop_cnt] 2. Test FIR config list, FIR amixer control id=$FIRid"
-        for config in ${FIRList[@]}; do
-            func_test_eq $FIRid $current_dir/$config
-            result=$?
-            if [[ $result -ne 0 ]]; then
-                dloge "Failed at $config"
-                let failed_cnt++
-            fi
-        done
-        dlogi "EQ test done: failed_cnt is $failed_cnt"
-        if [ $failed_cnt -gt 0 ]; then
+    # if FIR filter is avilable, test wit FIR coef
+    is_fir=$(echo $eq_support | grep -i fir | wc -l)
+    if [[ $is_fir -ne "" ]]; then
+        # 2. Test FIR
+        dlogi "2.1 Get amixer control id for FIR"
+        FIRid=`amixer -D hw:$sofcard controls | grep EQFIR| head -1| sed 's/numid=\([0-9]*\),.*/\1/'`
+        if [ -z $FIRid ]; then
+                dloge "can't find FIR filter"
+                exit 1
+        fi
+
+        declare -a FIRList=($(ls -d ${current_dir}/eqctl/eq_fir_*.txt))
+        nFIRList=${#FIRList[@]}
+        dlogi "2.2 FIR list $nFIRList ${FIRList[*]}"
+        if [ $nFIRList ==  0 ]; then
+            dloge "FIR flter coeff list error!"
             exit 1
         fi
-    done
+
+        for i in $(seq 1 $loop_cnt)
+        do
+            dlogi "[$i/$loop_cnt] 2.3 Test FIR config list, FIR amixer control id=$FIRid"
+            for config in ${FIRList[@]}; do
+                func_test_eq $FIRid $current_dir/$config
+                result=$?
+                if [[ $result -ne 0 ]]; then
```

Comment:
Shorter and compatible with `set -e`:
```shell
func_test_eq "$FIRid" "$current_dir/$config" || {
    dloge "Failed at $config"
    : $((failed_cnt++))
}
```

Reply:
Yes shorter but I like current style better, as it is more intuitive.

Reply:
Sorry but testing `$?` it not acceptable for a number of reasons: https://github.com/koalaman/shellcheck/wiki/SC2181
The main reason is probably compatibility with `set -e`, see other reasons above.

The `do_something || handle_error` style is very common and reads like plain English, you can get used to it quickly. If it's really too hard you can try this instead:

```shell
if ! func_test_eq "$FIRid" "$current_dir/$config"; then
    dloge "Failed at $config"
    : $((failed_cnt++))
fi
```

Reply:
In some cases you need to save the error for later; that is the real purpose of `$?`:
```shell
set -e
saved_error=0
do_something || saved_error=$? # saved _for later_
do_something_else1
do_something_else2
if [ 0 -ne $saved_error] then
...
```
But without any `do_something_else` then`$?` is at best pointless and at worst harmful.